Framework: Use a more functional programming approach to remove headers returned by the API

stephanethomas opened the pull request:
This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/3844. It updates code that removes headers returned by the API to use a [more functional programming approach](https://github.com/Automattic/wp-calypso/pull/3844#discussion_r55972607). To be more specific it replaces:

```javascript
delete data._headers
```

With:

```javascript
omit( data, '_headers' );
```

#### Testing instructions
 
1. Run `git checkout fix/delete-headers` and start your server
2. Get a test account and blog
3. Enable the store sandbox
4. Open the [`Stats` page](http://calypso.localhost:3000/stats/insights) and check that it displays as before
5. Open the [`Plans` page](http://calypso.localhost:3000/plans) and check that it displays as before
6. Open the [`Compare Plans` page](http://calypso.localhost:3000/plans/compare) and check that it displays as before
7. Navigate to the [`Domains` page](http://calypso.localhost:3000/domains/add)
8. Click on any `Add` button to add a domain to the shopping cart
9. Click on the `No thanks, I don't need email or will use another provider.` link on the next page
10. Check that the shopping cart is correct on the `Domain Contact Information` page 
11. Check that lists of countries and states display correctly on the `Domain Contact Information` page 
12. Fill in the form and click on the `Add Privacy Protection` button
13. Submit the form on the `Secure Payment` page with fake credit card information
14. Navigate to the [`Billing History` page](https://calypso.localhost:3000/me/billing)
15. Check that you can see your credit card at the bottom of the page
 
#### Reviews
 
- [x] Code
- [x] Product

drewblaisdell commented on client/lib/plans-list/index.js:
```diff
@@ -96,15 +97,13 @@ PlansList.prototype.initialize = function( plans ) {
-	delete data._headers;
-
-	return data;
+	return omit( data, '_headers' );
```
This endpoint returns an array, to which a `_headers` property is assigned:

![screen shot 2016-03-16 at 2 52 11 pm](https://cloud.githubusercontent.com/assets/1130674/13829902/bc329e5a-eb86-11e5-9e35-75375040cdce.png)

_.omit returns an object, but to preserve the behavior of this function, which currently returns an array, we should use _.reject here.

drewblaisdell commented on client/lib/products-list/index.js:
```diff
@@ -100,15 +101,13 @@ ProductsList.prototype.initialize = function( productsList ) {
-	delete data._headers;
-
-	return data;
+	return omit( data, '_headers' );
```
Unlike features/countries/plans-list, the products-list endpoint returns an object, so _.omit is appropriate here. :+1:

drewblaisdell commented:
I added a commit to use _.reject to remove `_headers` from responses that are an array instead of an object (otherwise, we break several different pages). If you're cool with that change, the code looks good to me.

scruffian commented:
Code :+1:

fabianapsimoes commented:
> Check that lists of countries and states display correctly on the Domain Contact Information page
> Fill in the form and click on the Add Privacy Protection button

After that, the credit card component never loads and I'm seeing this error in the console:

```TypeError: this.props.cards.get(...).map is not a function```

![screenshot-calypso localhost 3000 2016-03-17 15-17-01](https://cloud.githubusercontent.com/assets/1248436/13848721/63e55558-ec53-11e5-974f-eeafd29044c7.png)

For the record, I have a number of stored credit cards in this account. 

When no cards are stored, I still see the console error, but the credit card form is loaded. However, seems I can't submit payments anyway. After over a minute the Submitting Payment notice never went away, and I can see from our admin tool no payment was taken.

stephanethomas commented:
It looks like I've been careless, I'm sorry about that @drewblaisdell. I think I forgot to clear my local storage which is why I didn't encounter any issue when I tested it. I should probably avoid pushing a pull request when it's late and that I'm about to logout.

I've double check all endpoints and it's true that almost most of them return an array rather than an object. I've reverted the change in `stored-cards` which was causing the error mentioned by @fabianapsimoes.

I've rebased and squashed the commit history. This is ready for another review.

fabianapsimoes commented:
Product :+1:

scruffian commented:
Code :+1:
